aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,358 @@
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow(bucket, i);
+            if (&entry.key == &key) {
+                return &entry.value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_mut<K: drop, V>(table: &mut SmartTable<K, V>, key: K): &mut V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow_mut(&mut table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow_mut(bucket, i);
+            if (&entry.key == &key) {
+                return &mut entry.value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Returns true iff `table` contains an entry for `key`.
+    public fun contains<K: drop, V>(table: &SmartTable<K, V>, key: K): bool {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow(&table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow(bucket, i);
+            if (&entry.key == &key) {
+                return true
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    /// Remove from `table` and return the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun remove<K: copy + drop, V>(table: &mut SmartTable<K, V>, key: K): V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow_mut(&mut table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow(bucket, i);
+            if (&entry.key == &key) {
+                let Entry {hash:_, key:_, value} = vector::swap_remove(bucket, i);
+                table.size = table.size - 1;
+                return value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Returns the length of the table, i.e. the number of entries.
+    public fun length<K, V>(table: &SmartTable<K, V>): u64 {
+        table.size
+    }
+
+    /// Return the load factor of the hashtable.
+    public fun load_factor<K, V>(table: &SmartTable<K, V>): u64 {
+        table.size * 100 / (table.num_buckets * table.bucket_size)
+    }
+
+    /// Update `split_load_threshold`.
+    public fun update_split_load_threshold<K, V>(table: &mut SmartTable<K, V>, split_load_threshold: u8) {
```

Comment:
Hmmm. This seems like a potentially dangerous operation. It changes the threshold but doesn't re-structure existing buckets. This means that we'll have a mix of buckets that are above and below this threshold. What's the rationale here?

Reply:
The load factor is per-table but not per bucket. It is \frac{total_num_entries}{num_buckets * target_bucket_size}. Thus it should be okay to change the threshold at any time. It just make splitting more aggressive or more conservative.
Agree?

Reply:
Makes sense. Thanks for explaining